### DISCUSSION TITLE
Add a API to change generated class name

### DISCUSSION THREAD

**Discussion by maxkatz6** [devlooped/ThisAssembly#71] [Ideas]
Right now "ThisAssembly" class is generated.
It works terribly with Rider, see https://youtrack.jetbrains.com/issue/RIDER-25136, when there is another internal ThisAssembly somewhere in other assembly that is not accessible. Rider tries to read another ThisAssembly type instead of using auto generated file.
While it's true, that this issue is in Rider, and should be fixed there, users of this source generator will be blocked for a while, when they use Rider (which is the only one option on Linux and only one good option on macOS).
Also I can imagine situation when developer wants to change type name due of other possible conflicts.

I see at least two ways to workaround it:
1. Add msbuild property to completely change ThisAssembly static class name.
2.  Add msbuild property to add a prefix to the ThisAssembly static class name. In this way brand of this project will be preserved, if it's important.

**Comment by rdeago** [comment]
[Nerdbank.GitVersioning](https://github.com/dotnet/Nerdbank.GitVersioning/blob/main/doc/dotnet.md) also generates a `ThisAssembly` class. Currently it would just add more constants if used together with `ThisAssembly`, but this compatibility is more of a coincidence than a feature.

**Reply by kzu** [reply]
The linked jetbrains issue seems to have been fixed since, though:

![image](https://user-images.githubusercontent.com/169707/170141813-8f6b5fc0-8ebf-4f42-b11f-a44cab16350c.png)

If so, I think this should be very low pri.

**Reply by maxkatz6** [reply]
Yes, it's not a blocker with Rider now. But there are still two points to have it customizable:
1. Possible conflicts with other source generators
2. Personally for me, I would want to use ThisAssembly.Strings generator, referencing these strings from XAML like so:
```xaml
<Button Content="{x:Static res:Buttons.Ok}" />
```
but instead ThisAssembly would force me to use:
```xaml
<Button Content="{x:Static res:ThisAssembly.Strings.Buttons.Ok}" />
```

**Reply by maxkatz6** [reply]
Though, to be fair, for XAML it's more work than just rename "ThisAssembly" class, but update whole strings generator

**Comment by kzu** [comment]
This is very low pri, since other source generators that use ThisAssembly (i.e. Nerbank.GitVersioning) also generates it as partial and so is 100% compatible.
